sorc/ncep_post.fd/UPP_PHYSICS.f

```diff
@@ -4501,6 +4504,473 @@ END SUBROUTINE CALSLR_UUTAH
+            im1 = iw(i)
+            if(VWND(ip1,J)==SPVAL .or. VWND(im1,J)==SPVAL .or. &
+               UWND(I,J-1)==SPVAL .or. UWND(I,J+1)==SPVAL) cycle
+            ABSV(I,J)   = ((VWND(ip1,J)-VWND(im1,J))*wrk2(i,j)                 &
+     &                    -  (UWND(I,J-1)*COSL(I,J-1)                          &
+                          -   UWND(I,J+1)*COSL(I,J+1))*wrk3(i,j)) * wrk1(i,j)  &
+                          + F(I,J)
+            DIV(I,J)    = ((UWND(ip1,J)-UWND(im1,J))*wrk2(i,j)                 &
+     &                    +  (VWND(I,J-1)*COSL(I,J-1)                          &
+                          -   VWND(I,J+1)*COSL(I,J+1))*wrk3(i,j)) * wrk1(i,j)  
+          ENDDO
+        END IF
+        if (npass > 0) then
+          do i=ista,iend
+            tx1(i) = absv(i,j)
+          enddo
+          do nn=1,npass
+            do i=ista,iend
+              tx2(i+1) = tx1(i)
+            enddo
+            tx2(1)    = tx2(im+1)
+            tx2(im+2) = tx2(2)
+            do i=2,im+1
+              tx1(i-1) = 0.25 * (tx2(i-1) + tx2(i+1)) + 0.5*tx2(i)
+            enddo
+          enddo
+          do i=ista,iend
+            absv(i,j) = tx1(i)
+          enddo
+        endif
+      END DO                               ! end of J loop
+
+!     deallocate (wrk1, wrk2, wrk3, cosl)
+! GFS use lon avg as one scaler value for pole point
+!
+      ! call poleavg(IM,JM,JSTA,JEND,SMALL,COSL(1,jsta),SPVAL,ABSV(1,jsta))
+!
+      call exch(absv(ista_2l:iend_2u,jsta_2l:jend_2u))
+      call fullpole(absv(ista_2l:iend_2u,jsta_2l:jend_2u),avpoles)     
+!
+      cosltemp=spval
+      if(jsta== 1) cosltemp(1:im, 1)=coslpoles(1:im,1)
+      if(jend==jm) cosltemp(1:im,jm)=coslpoles(1:im,2)
+      avtemp=spval
+      if(jsta== 1) avtemp(1:im, 1)=avpoles(1:im,1)
+      if(jend==jm) avtemp(1:im,jm)=avpoles(1:im,2)
+!        
+      call poleavg(IM,JM,JSTA,JEND,SMALL,cosltemp(1,jsta),SPVAL,avtemp(1,jsta))
+!
+      if(jsta== 1) absv(ista:iend, 1)=avtemp(ista:iend, 1)
+      if(jend==jm) absv(ista:iend,jm)=avtemp(ista:iend,jm)
+!    
+!        deallocate (wrk1, wrk11, wrk2, wrk3, cosl, iw, ie)
+!    
+      call exch(absv(ista_2l:iend_2u,jsta_2l:jend_2u))
+      call exch(div(ista_2l:iend_2u,jsta_2l:jend_2u))
+!
+! store absv and div factors before poisson loops
+!$omp parallel do private(i,j)
+      do j=jsta,jend
+      do i=ista,iend
+        atmp(i,j)=0.25*(absv(i,j)-f(i,j))/(wrk2(i,j)*wrk1(i,j)*wrk3(i,j)*wrk1(i,j)*cosl(i,j)*4.)
+        dtmp(i,j)=0.25*div(i,j)/(wrk2(i,j)*wrk1(i,j)*wrk3(i,j)*wrk1(i,j)*cosl(i,j)*4.)
+      enddo
+      enddo
+!
+! poisson solver for psi and chi 
+      psi=0.
+      ta=mpi_wtime()
+      do jjk=1,1000
+      do jj=1,300 
+        call exch(psi(ista_2l:iend_2u,jsta_2l:jend_2u))
+        ptmp=psi
+        do j=jsta,jend
+        do i=ista,iend
+          if (j>1 .and. j<jm) then
+            psi(i,j) = 0.25*(ptmp(i-1,j)+ptmp(i+1,j)+ptmp(i,j-1)+ptmp(i,j+1))-atmp(i,j)
+            edif=psi(i,j)-pval
```

Comment:
@KarinaAsmar-NOAA @JesseMeng-NOAA What is the 'pval' used for? I don't see it initialized or calculated?

Reply:
This is to evaluate the differences between the previous and current iteration so should be
edif=psi(i,j)-ptmp(i,j)